prov/sm2/src/sm2_atomic.c

```diff
@@ -0,0 +1,548 @@
+		break;
+	case ofi_op_atomic_compare:
+		xfer_entry->hdr.size = ofi_copy_from_iov(
+			atomic_entry->atomic_data.buf,
+			SM2_ATOMIC_COMP_INJECT_SIZE, iov, iov_count, 0);
+		comp_size = ofi_copy_from_iov(atomic_entry->atomic_data.comp +
+						      xfer_entry->hdr.size,
+					      SM2_ATOMIC_COMP_INJECT_SIZE,
+					      compare_iov, compare_count, 0);
+		if (comp_size != xfer_entry->hdr.size)
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"atomic and compare buffer size mismatch\n");
+		break;
+	default:
+		assert(0);
+		break;
+	}
+}
+
+static ssize_t
+sm2_do_atomic_inject(struct sm2_ep *ep, int64_t peer_gid, uint32_t op,
+		     uint64_t op_flags, uint8_t datatype, uint8_t atomic_op,
+		     const struct iovec *iov, size_t iov_count,
+		     const struct fi_rma_ioc *rma_ioc, size_t rma_ioc_count,
+		     const struct iovec *compare_iov, size_t compare_count,
+		     const struct iovec *result_iov, size_t result_count,
+		     size_t total_len, void *context)
+{
+	struct sm2_xfer_entry *xfer_entry;
+	size_t ret;
+
+	ret = sm2_pop_xfer_entry(ep, &xfer_entry);
+	if (ret)
+		return ret;
+
+	xfer_entry->hdr.proto = sm2_proto_inject;
+
+	sm2_generic_format(xfer_entry, ep->gid, op, 0, 0, op_flags, context);
+	sm2_format_atomic_op(xfer_entry, datatype, atomic_op);
+	sm2_format_atomic_rma_ioc(xfer_entry, rma_ioc, rma_ioc_count);
+	sm2_format_atomic_result(xfer_entry, result_iov, result_count);
+	sm2_format_atomic_data(xfer_entry, iov, iov_count, compare_iov,
+			       compare_count);
+	sm2_fifo_write(ep, peer_gid, xfer_entry);
+
+	return FI_SUCCESS;
+}
+
+static ssize_t sm2_generic_atomic(
+	struct sm2_ep *ep, const struct fi_ioc *ioc, void **desc,
+	size_t iov_count, const struct fi_ioc *compare_ioc, void **compare_desc,
+	size_t compare_count, struct fi_ioc *result_ioc, void **result_desc,
+	size_t result_count, fi_addr_t addr, const struct fi_rma_ioc *rma_ioc,
+	size_t rma_ioc_count, enum fi_datatype datatype, enum fi_op atomic_op,
+	void *context, uint32_t op, uint64_t op_flags)
+{
+	struct iovec iov[SM2_IOV_LIMIT];
+	struct iovec compare_iov[SM2_IOV_LIMIT];
+	struct iovec result_iov[SM2_IOV_LIMIT];
+	sm2_gid_t peer_gid;
+	ssize_t ret = 0;
+	size_t total_len;
+
+	assert(iov_count <= SM2_IOV_LIMIT);
+	assert(result_count <= SM2_IOV_LIMIT);
+	assert(compare_count <= SM2_IOV_LIMIT);
+	assert(rma_ioc_count <= SM2_IOV_LIMIT);
+
+	ret = sm2_verify_peer(ep, addr, &peer_gid);
+	if (ret < 0)
+		return ret;
+
+	/* TODO locking?... */
+
+	total_len =
+		ofi_datatype_size(datatype) * ofi_total_ioc_cnt(ioc, iov_count);
+
+	switch (op) {
+	case ofi_op_atomic_compare:
+		assert(compare_ioc);
+		ofi_ioc_to_iov(compare_ioc, compare_iov, compare_count,
+			       ofi_datatype_size(datatype));
+		total_len *= 2;
+		/* fall through */
+	case ofi_op_atomic_fetch:
+		assert(result_ioc);
+		ofi_ioc_to_iov(result_ioc, result_iov, result_count,
+			       ofi_datatype_size(datatype));
+		op_flags |= SM2_RMA_REQ;
+		op_flags |= FI_DELIVERY_COMPLETE;
```

Comment:
I can't remember if we've already talked about this but I don't think it makes sense to use the op_flags to hold both sm2 flags and fi flags. In shm we have them as separate. I know you've simplified it to remove the need for the sm2 specific flags which I think is a good direction (I never liked the translation between fi and shm flags) but now you're bringing it back in and using them at the same time. Maybe use FI_REMOTE_READ instead?

Reply:
Done, this might have implications to @lrbison rma PR.

Reply:
I agree with Alexia here.  I tried to keep my SM2-specific flags for RMA/SAR separate so I don't think this will impact me.

I think it may be worth while to have 'protocol_flags' in the xfer_entry header, where we can put SM2 specific things.